qiskit/transpiler/passes/optimization/elide_permutations.py

```diff
@@ -0,0 +1,106 @@
+        of the circuit as input to this pass.
+    * ``virtual_permutation_layout``: A :class:`~.Layout` object mapping input qubits to the output
+        state after eliding permutations.
+
+    These three properties are needed for the transpiler to track the permutations in the out
+    :attr:`.QuantumCircuit.layout` attribute. The elision of permutations is equivalent to a
+    ``final_layout`` set by routing and all three of these attributes are needed in the case
+    """
+
+    def run(self, dag):
+        """Run the ElidePermutations pass on ``dag``.
+
+        Args:
+            dag (DAGCircuit): the DAG to be optimized.
+
+        Returns:
+            DAGCircuit: the optimized DAG.
+        """
+        if self.property_set["layout"] is not None:
+            logger.warning(
+                "ElidePermutations is not valid after a layout has been set. This indicates "
+                "an invalid pass manager construction."
+            )
+            return dag
+
+        op_count = dag.count_ops()
+        if op_count.get("swap", 0) == 0 and op_count.get("permutation", 0) == 0:
+            return dag
+
+        new_dag = dag.copy_empty_like()
+        qubit_mapping = list(range(len(dag.qubits)))
+
+        def _apply_mapping(qargs):
+            return tuple(dag.qubits[qubit_mapping[dag.find_bit(qubit).index]] for qubit in qargs)
+
+        for node in dag.topological_op_nodes():
+            if not isinstance(node.op, (SwapGate, PermutationGate)):
+                new_dag.apply_operation_back(node.op, _apply_mapping(node.qargs), node.cargs)
+            elif getattr(node.op, "condition", None) is not None:
+                new_dag.apply_operation_back(node.op, _apply_mapping(node.qargs), node.cargs)
+            elif isinstance(node.op, SwapGate):
+                index_0 = dag.find_bit(node.qargs[0]).index
+                index_1 = dag.find_bit(node.qargs[1]).index
+                qubit_mapping[index_1], qubit_mapping[index_0] = (
+                    qubit_mapping[index_0],
+                    qubit_mapping[index_1],
+                )
+            elif isinstance(node.op, PermutationGate):
+                starting_indices = [qubit_mapping[dag.find_bit(qarg).index] for qarg in node.qargs]
+                pattern = node.op.params[0]
+                pattern_indices = [qubit_mapping[idx] for idx in pattern]
+                for i, j in zip(starting_indices, pattern_indices):
+                    qubit_mapping[i] = j
+        input_qubit_mapping = {qubit: index for index, qubit in enumerate(dag.qubits)}
+        self.property_set["original_layout"] = Layout(input_qubit_mapping)
+        if self.property_set["original_qubit_indices"] is None:
+            self.property_set["original_qubit_indices"] = input_qubit_mapping
+        # ToDo: check if this exists; then compose
+        self.property_set["virtual_permutation_layout"] = Layout(
+            {dag.qubits[out]: idx for idx, out in enumerate(qubit_mapping)}
+        )
```

Comment:
I'd like to review properly once I'm back at work on Tuesday, but just as an immediate note: please can we consider not using `Layout` here?  As both its name and the top PR comment says, the `virtual_permutation` is a permutation of virtual-to-virtual, whereas the `Layout` class is explicitly a mapping of virtual to physical qubits (i.e. `Layout` is a map from one Hilbert space to a different Hilbert space, whereas this is an operator within the same Hilbert space).  The `idx` here are not physical qubits, they're just virtual-qubit indices, and I think we ought to separate the representations, especially given how confused we keep getting about `Layout`.

Reply:
By way of alternative, I mean: I'd prefer just to store an int list of the permutation, which is how we store permutations in all other places (like `PermutationGate`).

Reply:
Jake, I very much agree with you that using the word _layout_ to represent anything else than _virtual-to-physical_ is confusing. For instance, it took me a lot of time and an offline discussion with Matthew to understand that _"final_layout"_ stands for the _physical-to-physical_ permutation added during the routing stage. Having said that, I feel that at this point having both _virtual-to-physical_ and _physical-to-physical_ maps called _layouts_ and internally represented using the ``Layout`` class, while having the _virtual-to-virtual_ map being something else would be even more confusing. So not my favorite choice, but I am arguing for keeping the _virtual-to-virtual_ map as a ``Layout`` as well.

It might be simplest to think of the word _layout_ as of some mapping $f: A \rightarrow B$, where $A$ and $B$ can be anything. One nice thing is that you can easily compose mappings, e.g. composing the mappings $f:A\rightarrow B$ and $g:B\rightarrow C$ produces the mapping $g\circ f:A\rightarrow C$. The functionality to compose ``Layout``s was recently implemented in PR #11399, and this simplifies the code that deals with ``Layout``s by a lot. 

In addition, we are now talking about multiple passes that can modify the _virtual-to-virtual_ permutation: e.g., this `ElidePermutations` pass and the new `StarPreRouting` pass. To keep track of the combined permutation we can simply compose ``Layouts``.

Reply:
I think storing `final_layout` as a `Layout` was a mistake, and that was borne out in how nobody was able to use/understand `TranspileLayout` at first (both us - see all the issues we had in implementing `Operator.from_circuit` - and users) and we had to completely remake the public interface in #10835.

In an ideal world, we'd walk back `final_layout` but we only didn't in #10835 because it would have been a breaking change so we just left it. I'd really prefer that we didn't expand that bad situation. I wouldn't say we represent physical-to-physical permutations everywhere as Layouts at all, we just do it in this one place that imo we already recognised as a mistake.

I'm also not super wild that we merged #11399 - to me, that PR has a mistake even in its top comment where is even suggests that a `Layout` can have an inverse.  A _permutation_ can have an inverse, but a layout can't - a Layout doesn't have a "forwards" and "backwards", it has a "virtual" side and a "physical" side, and every part of the data structure, public method and argument typing says that.

Reply:
I am fine with any decision but I do find it so much easier to think about all of this (including how to guarantee that unitary equivalence is preserved) when I think of a _layout_ as of an _invertible mapping_. For better or for worse, even the "layout" (in the _property set_ terminology) aka the "initial_layout" (in `TranspileLayout` terminilogy) is an invertible mapping from ancilla-expanded virtual qubits to physical qubits. I would argue that this is also a mistake, but this is another thing that we are stuck with for now. But if we do work with invertible mappings, having the inverse mapping is very natural and really helps writing shorter cleaner code. At this point I feel it would be significantly more confusing if we did not repeat the same mistake once again. :)

Reply:
How possible do you think to introduce `TranspileLayoutV2`, and have only it in the passmanager's property set (instead of the "layout", "final_layout", "virtual_permutation_layout" and so on)?

Reply:
> I think of a _layout_ as of an _invertible mapping_.

This is totally fair for some mathematical concept of a layout as a single function, but imo not right for Qiskit's `Layout` class.  `Layout` is not invertible, because it's already a bijective, explicitly bidirectional map; it is self-inverse.  Its `__getitem__` signature is
```python
class Layout[A, B]:
    @overload
    def __getitem__(self, in_: A) -> B: ...
    @overload
    def __getitem__(self, out: B) -> A: ...
```
where `A` and `B` are virtual qubits (using `Qubit`) and physical qubits (using `int`) respectively.  Its image and domain are both `Union[A, B]` for valid `A` and `B`, or `set.union(Layout._v2p.keys(), Layout._p2v.keys())` if you prefer.  In the mathematical terminology we've used elsewhere of $L: V\to P$ and $L^{-1}: P\to V$, the `__getitem__` of Qiskit's `Layout` is simultaneously $L$ _and_ $L^{-1}$.  Trying to let it represent permutations where the types `A` and `B` are actually equal (which the class was never designed to permit) is why we suddenly need new arguments everywhere to `inverse`, `compose` and `to_permutation`, to _override_ the existing image and domain (both are `layout._v2p.keys() | layout._p2v.keys()` - again, note how it's baked in that virtual and physical are different spaces) of the class mapping.

We already have a way of representing permutations in Qiskit, and imo we'd be much more consistent with ourselves and much easier to move things like implicit permutations into explicit ones if we use the same conventions as our existing `PermutationGate` to represent permutations within the same Hilbert space.  The logic for `inverse` and `compose` for "list of int" (in the sense of "qubit `perm[i]` moves to index `i`") is very simple:

```python
def inverse(perm):
    reverse = {in_: out for out, in_ in enumerate(perm)}
    return [reverse[in_] for in_ in range(len(reverse))]

def compose(*perms):
    if not perms:
        return []
    out = range(len(perms[0]))
    for perm in perms:
        out = [perm[i] for i in out]
    return out
```

If working with the integer lists directly is unpleasant, my preferred form would be to add a new `Permutation` class to `qiskit/transpiler` that is basically just a list of ints, but implements `inverse` and `compose`.

The advantages of my proposal are, imo:

1. we avoid overloading `Layout` to mean things other than the $V \leftrightarrow P$ mapping that it explicitly is.
2. there's a clear, very simple way to explain the logic of things like `Operator.from_circuit(qc)`: it becomes `Operator(qc.append(PermutationGate(qc.layout.routing_permutation().inverse()), qc.qubits))`.
3. Similarly, `ElideSwaps` can now also be understood as just pushing all the swaps to the end of the circuit into a `PermutationGate`, then we just take that `PermutationGate` off the circuit and stick its `PermutationGate.pattern.inverse()` verbatim into `property_set["final_permutation"]` or the like.

Both of points 2 and 3 become way harder to understand when we have to indirect through a `Layout` object that was simply not intended to represent them.

---

Fwiw, I showed how the `TranspileLayout` class already generated everything that was needed for `Operator.from_circuit` in easy-to-use forms and how it integrates with the existing `PermutationGate` conventions over in https://github.com/Qiskit/qiskit/pull/12057#discussion_r1546515861 (note that it _does not_ use the mistaken `final_layout` attribute nor `routing_permutation`, neither of which are especially useful in that case).  I don't believe we need the #11399 methods on `Layout` to achieve those, since `TranspileLayout` was directly meant to replace them.  The only thing that I think we would want anew is the `compose` method I put above, but I think I've argued enough that I don't think that's a `Layout` method, it's a `PermutationGate` (or a new `Permutation`, if we want it to be lighter weight) method.

Reply:
A quick reply for now. These are very good points. I agree that working with `TranspileLayout` class is cleaner than working with different `Layout`s (which are not really layouts), but who and when sets `circuit._layout` as in `transpiled.layout`? I saw that different passes set "layout" and "final_layout" attributes in the property set, but I did not understand when "circuit._layout" is updated. The math for "undoing" the transpiled circuit is pretty much the same as in your comment https://github.com/Qiskit/qiskit/pull/12057#discussion_r1546515861 and what's currently implemented in #12057; in terms of the implementation, please note that we already have function `operator.apply_permutation` that are equivalent to applying a permutation gate before or after the operator and are probably a bit faster (though I did not check this). Regarding the point (3) on `ElideSwaps`, I don't think we can directly update the "final_layout" attribute, I believe the math here https://github.com/Qiskit/qiskit/pull/9523#issue-1568479275 is correct, and for this we need to know the "layout" which is not yet available at the time of `ElidePermutations`; this is why I am suggesting to save it on the side and update the final layout at the very end in a separate pass, but I would be more than happy to see alternative solutions.

Reply:
`TranspileLayout` is calculated at the end of `PassManager.run`. We can change how it's calculated as we choose - you can derive the components from the methods and vice versa. To avoid working with the very tricksy and misleading `final_layout`, we could easily add an alternative constructor to `TranspileLayout` that takes in a proper initial layout and a proper final layout (as in, actual layouts) and the virtual-qubit order and calculate back from that, or really any other set of objects.

Your concern here, iiuc, is that `TranspileLayout` isn't available in transpiler passes until the end of the pipeline. That's a big part of my reason for absolutely not wanting to reinvent it on `Layout`, which is effectively what #11399 does, and instead to represent the intermediate states within the transpiler pipeline as true permutations.

I'm sure your maths in the top is correct (I haven't checked), but its's complicated by the fact it keeps indirecting itself through a `Layout`, when it doesn't need to if we don't try and write everything in terms of `Layout`, which is unsuited to the task.

Reply:
So, would it be possible to have `TranspileLayout` available to every transpiler pass (and not just at the end of the run) with every pass being able to update the fields of the `TranspileLayout` and not the "layout" and "final_layout" attributes of the property set? (And if these are needed for backward compatibility, then somehow retrieve them from the `TranspileLayout`)?

Oh yeah the whole reason to invent the `FinalizeTranspiler` pass was that `SetLayout` and `ApplyLayout` do not always run (iirc they do not run when no coupling map or target is specified). I thought there is a deep reason why we might want to skip running these passes, but if not, we could indeed delegate the job of lifting the permutation of virtual qubits to a permutation of physical qubits to the `ApplyLayout` pass (which was actually Matthew's original suggestion).

Reply:
Thanks Jake, I am now very much in favor of the changes that you are proposing. If we think of the "final_layout" **not** as of a layout (i.e. not as of a map $V\to P$) and not as of a routing permutation of physical qubits (i.e. not as of a map $P\to P$), but as representing a permutation that is implicitly applied at the end of the current circuit, then everything can be realized just you have described and we don't need to explicitly store the "virtual_permutation_layout" or introduce the `FinalizeLayouts` pass. Great! This is actually how I was thinking about what's going on all along (i.e. a circuit with an implicit permutation gate at the end), but then tried to fit this into how I understood the "layout" / "final layout" framework.

However, in order to avoid the same confusion in the future, can we actually replace "final layout" from `Layout` class by an actual permutation  (both choices as a list of integers e.g. `[2, 3, 1, 0]` or as a permutation gate `PermutationGate([2, 3, 1, 0])` look fine for me) and not as of an object of type `Layout`? I am afraid that technically this change may not be backward-compatible, however these are somewhat of internal transpiler objects, and in theory the user should only be working with the final `TranspileLayout`. What do you think?

On the topic, where in Qiskit do we document what each property set attribute really means? For instance, I don't see anything relevant when searching for `final_layout`. 

_Aside:_ when talking about permutations we have two different conventions going on. IIRC, a `routing_permutation` in `TranspileLayout` of the form `[2, 3, 1, 0]` means `0 -> 2, 1 -> 3, 2 -> 1, 3 -> 0` while for `PermutationGate([2, 3, 1, 0])` the convention is `2->0, 2->1, 1->2, 0->3`. The two are inverse of each other, but it's still annoying to recall which one is used when.

I completely agree on the details of composing permutation and applying permutations to layouts. However, I want a common place that implements these methods, rather than having many passes duplicating the same code. This is why I was hoping that the passes could work with `TranspileLayout` class which already has (or should have) relevant usability functions.

I think we also agree that every pass implicitly takes the triple `(current dag, current layout, current implicit permutation)` and is responsible for producing the triple `(new dag, new layout, new implicit permutation)` which should be unitary-equivalent (using the `Operator.from_circuit` logic) provided that the circuit only consists of unitary operations and there is no ancilla expansion.

Back to the backwards-compatibility question and documentation, we would not be telling the writers of say routing passes (that can be written as external routing plugins) to make sure to update the "layout" and "final_layout" attributes. Is this backward-compatible? Where do or should we document this?

Reply:
For backwards compatibility, I think something like this could work:
- make `final_permutation` a different property to `final_layout` with the type we want it to have
- have Qiskit-internal passes _not_ set `final_layout`, but only set `final_permutation`
- during `TranspileLayout` construction at the end of the circuit, _if_ `final_layout` is set, then assume that it ran during a user-defined routing stage and compose its implied permutation with `final_permutation`.  We may well get that wrong if we ran a swap-elision pass _after_ a user routing pass, but I think that's very unlikely.
- routing and elision passes would never update `layout` (since that's just the _initial_ layout), only `final_permutation`. 

For different conventions between `routing_permutation` and `PermutationGate`: that's part of what I was getting at with "we can either store the inverse or the forwards permutation".  You could argue that they're the same permutation, it's just one is stored as an additive (apply this to a circuit to undo the permutation) and one is subtractive (apply the inverse of this to a circuit to undo the permutation).  We can make `final_permutation` store whichever of those matches the convention of `routing_permutation` already - as you say, you can easily convert one to the other - so we're consistent with ourselves.